File: go/day2/main.go
package main

import (
	"fmt"
	"github.com/vncsalencar/advent-of-code/util"
	"strings"
)

func main() {
	input := util.GetInput()
	lines := strings.Split(input, "\n")

	var part1 int32 = 0
	var part2 int32 = 0
	// A      B       C
	// X      Y       Z
	// rock paper   scissors
	moves1 := map[string]int32{
		"A X": 1 + 3,
		"A Y": 2 + 6,
		"A Z": 3 + 0,
		"B X": 1 + 0,
		"B Y": 2 + 3,
		"B Z": 3 + 6,
		"C X": 1 + 6,
		"C Y": 2 + 0,
		"C Z": 3 + 3,
	}

	moves2 := map[string]int32{
		"A X": 3 + 0,
		"A Y": 1 + 3,
		"A Z": 2 + 6,
		"B X": 1 + 0,
		"B Y": 2 + 3,
		"B Z": 3 + 6,
		"C X": 2 + 0,
		"C Y": 3 + 3,
		"C Z": 1 + 6,
	}

	for _, line := range lines {
		part1 += moves1[line]
		part2 += moves2[line]
	}

	fmt.Println("# Day 2: Rock Paper Scissors")
	fmt.Println("Part 1 =", part1)
	fmt.Println("Part 2 =", part2)
}

File: rust/day2/src/main.rs
use std::{error, str::FromStr};

fn main() -> Result<(), Box<dyn error::Error>> {
    let input = include_str!("./input.txt");

    let part1 = input
        .lines()
        .map(|line| {
            let line = line.split(" ").collect::<Vec<&str>>();
            let opponent = line[0].parse::<Movement>().expect("Invalid movement");
            let me = line[1].parse::<Movement>().expect("Invalid movement");
            me.points() + me.duel(&opponent)
        })
        .sum::<usize>();

    let part2 = input
        .lines()
        .map(|line| {
            let line = line.split(" ").collect::<Vec<&str>>();
            let opponent = line[0].parse::<Movement>().expect("Invalid movement");
            let me = Movement::from_desired_outcome(&opponent, line[1]);
            me.points() + me.duel(&opponent)
        })
        .sum::<usize>();

    println!("# Day 2: Rock Paper Scissors");
    println!("Part 1 = {part1}");
    println!("Part 2 = {part2}");
    Ok(())
}

#[derive(Clone, PartialEq)]
enum Movement {
    Paper,
    Rock,
    Scissors,
}

impl FromStr for Movement {
    type Err = Box<dyn error::Error>;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "A" | "X" => Ok(Self::Rock),
            "B" | "Y" => Ok(Self::Paper),
            "C" | "Z" => Ok(Self::Scissors),
            _ => panic!("Invalid move value"),
        }
    }
}

impl Movement {
    fn from_desired_outcome(opponent: &Self, outcome: &str) -> Self {
        return match (opponent, outcome) {
            (_, "X") => opponent.wins_against(),
            (_, "Y") => opponent.clone(),
            (_, "Z") => opponent.loses_against(),
            _ => panic!("Invalid movement or outcome value"),
        };
    }

    fn wins_against(&self) -> Self {
        match self {
            Self::Rock => Self::Scissors,
            Self::Paper => Self::Rock,
            Self::Scissors => Self::Paper,
        }
    }

    fn loses_against(&self) -> Self {
        match self {
            Self::Rock => Self::Paper,
            Self::Paper => Self::Scissors,
            Self::Scissors => Self::Rock,
        }
    }

    fn points(&self) -> usize {
        match self {
            Self::Rock => 1,
            Self::Paper => 2,
            Self::Scissors => 3,
        }
    }

    fn duel(&self, other: &Self) -> usize {
        match (self, other) {
            (Self::Rock, Self::Scissors)
            | (Self::Paper, Self::Rock)
            | (Self::Scissors, Self::Paper) => 6,
            _ if other == self => 3,
            _ => 0,
        }
    }
}
